dbms/src/Interpreters/ActionsVisitor.cpp

```diff
@@ -123,13 +123,41 @@ SetPtr makeExplicitSet(
     else if (left_tuple_depth + 1 == right_tuple_depth)
     {
         const auto * set_func = right_arg->as<ASTFunction>();
-
-        if (!set_func || set_func->name != "tuple")
+        if (!set_func)
             throw Exception("Incorrect type of 2nd argument for function " + node->name
-                            + ". Must be subquery or set of elements with type " + left_arg_type->getName() + ".",
+                            + ". Must be subquery, set of elements or constant function with type " + left_arg_type->getName() + ".",
                             ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+        if (set_func->name == "tuple")
+        {
+            elements_ast = set_func->arguments;
+        }
+        else
+        {
+            // Evaluate a function and unpack a result into ASTExpressionList:
+            elements_ast = std::make_shared<ASTExpressionList>();
 
-        elements_ast = set_func->arguments;
+            auto result = evaluateConstantExpression(right_arg, context);
+            switch (result.first.getType())
+            {
+                // multiple values: make a list of ASTLiterals
+                case Field::Types::Array:
+                case Field::Types::Tuple:
+                {
+                    Array & array = DB::get<Array &>(result.first);
+                    for (auto & a : array)
+                    {
+                        elements_ast->children.push_back(std::make_shared<ASTLiteral>(std::move(a)));
```

Comment:
Slow. Can we use Block variant of Set creation instead of creating and destroying temporary useless AST?

Reply:
Fixed that